src/ngMessages/messages.js

```diff
@@ -228,91 +304,214 @@ angular.module('ngMessages', [])
+         var ctrl = this;
+         var latestKey = 0;
 
-        this.renderMessages = function(values, multiple) {
-          values = values || {};
+         var messages = this.messages = {};
+         var renderLater, cachedCollection;
 
-          var found;
-          angular.forEach(messages, function(message) {
-            if ((!found || multiple) && truthyVal(values[message.type])) {
-              message.attach();
-              found = true;
-            } else {
-              message.detach();
-            }
-          });
+         this.render = function(collection) {
+           collection = collection || {};
 
-          this.renderElementClasses(found);
+           renderLater = false;
+           cachedCollection = collection;
 
-          function truthyVal(value) {
-            return value !== null && value !== false && value;
-          }
-        };
-      },
-      require: 'ngMessages',
-      link: function($scope, element, $attrs, ctrl) {
-        ctrl.renderElementClasses = function(bool) {
-          bool ? $animate.setClass(element, ACTIVE_CLASS, INACTIVE_CLASS)
-               : $animate.setClass(element, INACTIVE_CLASS, ACTIVE_CLASS);
-        };
+           // this is true if the attribute is empty or if the attribute value is truthy
+           var showAll = isAttrTruthy($scope, $attrs.ngMessagesMultiple) ||
+                         isAttrTruthy($scope, $attrs.multiple);
 
-        //JavaScript treats empty strings as false, but ng-message-multiple by itself is an empty string
-        var multiple = angular.isString($attrs.ngMessagesMultiple) ||
-                       angular.isString($attrs.multiple);
+           var matches = [];
+           var nameMatches = {};
+           var item = ctrl.head;
+           var messageExists = false;
+           while (item != null && !messageExists) {
+             var message = item.message;
+             forEach(collection, function(value, key) {
+               if (!messageExists && truthy(value) && message.test(key)) {
+                 if (showAll) {
+                   // this is to prevent the same error name from showing up twice
+                   if (nameMatches[key]) return;
+                   nameMatches[key]=true;
+                 }
+                 matches.push(item);
+                 messageExists = !showAll;
+                 message.attach();
+               }
+             });
+             item = item.next;
+           }
 
-        var cachedValues, watchAttr = $attrs.ngMessages || $attrs['for']; //for is a reserved keyword
-        $scope.$watchCollection(watchAttr, function(values) {
-          cachedValues = values;
-          ctrl.renderMessages(values, multiple);
-        });
+           // all of the messages that were not true are not to be displayed
+           forEach(messages, function(item) {
+             if (matches.indexOf(item) == -1) {
+               item.message.detach();
+             }
+           });
```

Comment:
Isn't there some way we could incorporate this into the previous loops to save on this extra loop with multiple indexOf calls.